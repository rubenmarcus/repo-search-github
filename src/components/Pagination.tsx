import {
  Pagination,
  PaginationContent,
  PaginationEllipsis,
  PaginationItem,
  PaginationLink,
  PaginationNext,
  PaginationPrevious,
} from "@/components/ui/pagination";
import { PaginationProps } from "@/types/types";

export const PaginationComponent = ({
  pages,
  currentPage,
  handlePrevClick,
  handleNextClick,
  handlePageClick,
  startPage,
}: PaginationProps) => {
  //disabled={currentPage === 1}

  console.log(currentPage, "currentPage");
  console.log(pages, 'pages')

  return (
    <Pagination className="my-5">
      <PaginationContent>
        <PaginationItem>
          <PaginationPrevious href="#" onClick={handlePrevClick} />
        </PaginationItem>

        {pages.map((page) => (
          <PaginationItem onClick={() => handlePageClick(page)}>
            <PaginationLink
              className={
                currentPage == startPage + page - 1 ? "bg-white text-black" : ""
              }
              href="#"
            >
              {startPage + page - 1}
            </PaginationLink>{" "}
          </PaginationItem>
        ))}

        <PaginationItem>
          <PaginationEllipsis />
        </PaginationItem>
        <PaginationItem>
          <PaginationNext onClick={handleNextClick} href="#" />
        </PaginationItem>
      </PaginationContent>
    </Pagination>
  );
};
