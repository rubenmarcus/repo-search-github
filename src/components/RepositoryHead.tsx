import { Input } from "@/components/ui/input";
import { Button } from "@/components/ui/button";
import { HeadProps } from "@/types/types";
import { Skeleton } from "./ui/skeleton";

export const RepositoryHead = ({
  headProps,
}: {
  headProps: HeadProps;
}): JSX.Element => {
  const { handleSubmit, searchInput, handleInputChange, totalRepos, topic } =
    headProps;

  return (
    <>
      <h1 className="scroll-m-20 text-4xl font-extrabold tracking-tight lg:text-3xl w-full pt-10">
        Github Repository Search
      </h1>
      <h2> Search Github repositories by topic </h2>
      <div className="flex w-full my-5 gap-10 items-center">
        <h3> {topic}</h3>
        <div className="flex gap-2 items-center">
          <p>Total Repositories:{" "}</p>
          {totalRepos ?? <Skeleton className="w-[80px] h-[12px] " />}
        </div>
        <form
          onSubmit={handleSubmit}
          className="ml-auto flex"
          data-testid="search-form"
        >
          <Input
            type="text"
            placeholder="Search by topic..."
            value={searchInput}
            name="search"
            onChange={(e) => {
              handleInputChange(e.target.value);
            }}
            className="border-stone-400 mr-2"
          />
          <Button
            className="primary hover:bg-white hover:text-black"
            type="submit"
          >
            Search
          </Button>
        </form>
      </div>
    </>
  );
};
