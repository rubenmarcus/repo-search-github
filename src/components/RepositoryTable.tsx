import { Skeleton } from "@/components/ui/skeleton";
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from "@/components/ui/table";
import { SearchResponse } from "@/types/types";

const LoadingTable = () => {
  const skeletonRow = (
    <>
      <TableCell className="font-medium">
        <Skeleton className="w-[100px] h-[20px] " />
      </TableCell>
      <TableCell>
        <Skeleton className="w-[100px] h-[20px] " />
      </TableCell>
      <TableCell>
        <Skeleton className="w-[100px] h-[20px] " />
      </TableCell>
      <TableCell className="text-right">
        <Skeleton className="w-[100px] h-[20px] " />
      </TableCell>
      <TableCell className="text-right">
        <Skeleton className="w-[100px] h-[20px] " />
      </TableCell>
      <TableCell className="text-right">
        <Skeleton className="w-[100px] h-[20px] " />
      </TableCell>
    </>
  );

  const skeletonRows = Array.from({ length: 10 }, (_, index) => (
    <TableRow className="border-stone-700" key={`skeleton-row-${index}`}>
      {skeletonRow}
    </TableRow>
  ));

  return (
    <Table className="text-xs bg-[#333]">
      <TableHeader>
        <TableRow className="table-row-head border-stone-700">
          <TableHead className="w-[100px]  text-white">Name</TableHead>
          <TableHead className="text-white">Owner</TableHead>
          <TableHead className="text-white">Description</TableHead>
          <TableHead className="text-right  text-white">Stargazers</TableHead>
          <TableHead className="text-right  text-white">Watchers</TableHead>
          <TableHead className="text-right  text-white">Forks</TableHead>
        </TableRow>
      </TableHeader>
      <TableBody>{skeletonRows}</TableBody>
    </Table>
  );
};

const TableContainer = ({ data }: { data: SearchResponse }): JSX.Element => {
  return (
    <Table className="text-xs bg-[#333]">
      <TableHeader>
        <TableRow className="table-row-head border-stone-700 ">
          <TableHead className="w-[140px] text-white">Name</TableHead>
          <TableHead className="w-[100px] text-white">Owner</TableHead>
          <TableHead className="w-[210px] text-white">Description</TableHead>
          <TableHead className="text-right text-white">Stargazers</TableHead>
          <TableHead className="text-right text-white">Watchers</TableHead>
          <TableHead className="text-right text-white">Forks</TableHead>
        </TableRow>
      </TableHeader>
      <TableBody>
        {data?.search.edges.map(({ node }) => (
          <TableRow
            key={node.url}
            className="border-stone-700 table-row hover:bg-[#444]"
          >
            <TableCell className="font-medium">
              <a href={node.url} target="_blank">
                {" "}
                {node.name}
              </a>
            </TableCell>
            <TableCell>{node.owner.login}</TableCell>
            <TableCell className="text-wrap">{node.description}</TableCell>
            <TableCell className="text-right">
              {node.stargazers.totalCount}
            </TableCell>
            <TableCell className="text-right">
              {node.watchers.totalCount}
            </TableCell>
            <TableCell className="text-right">
              {node.forks.totalCount}
            </TableCell>
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );
};

export const RepositoryTable = ({ data, loading }: {data: SearchResponse | undefined, loading:boolean}): JSX.Element => {
  if (loading) {
    return <LoadingTable />;
  }

  return data ? <TableContainer data={data} /> : <></>;
};
