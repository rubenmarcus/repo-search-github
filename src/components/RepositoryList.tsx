import { useRepositoryData } from "../hooks/useReposData";
import { PaginationComponent } from "./Pagination";
import { RepositoryHead } from "./RepositoryHead";
import { RepositoryTable } from "./RepositoryTable";

const RepositoryList = () => {
  const { loading, error, data, headProps, paginationProps, noRepos } =
    useRepositoryData();


    console.log(JSON.stringify(headProps, undefined, 3))

  if (error) return <p>Error: {error.message}</p>;

  return (
    <div className="container h-full pb-20">
      <RepositoryHead {...{ headProps }} />

      {noRepos ? (
        <h1 className="scroll-m-20 text-4xl font-extrabold tracking-tight lg:text-2xl w-full pt-10">
          No Repositories Found, please try another topic.
        </h1>
      ) : (
        <>
          <RepositoryTable {...{ data, loading }} />
          <PaginationComponent {...paginationProps} />
        </>
      )}
    </div>
  );
};

export default RepositoryList;
