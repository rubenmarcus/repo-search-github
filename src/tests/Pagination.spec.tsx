import { render, fireEvent } from "@testing-library/react";
import { PaginationComponent } from "@/components/Pagination";

describe("PaginationComponent", () => {
  const mockPages = [1, 2, 3, 4, 5];
  const mockCurrentPage = 1;
  const mockHandlePrevClick = jest.fn();
  const mockHandleNextClick = jest.fn();
  const mockHandlePageClick = jest.fn();
  const mockStartPage = 1;

  it("renders the PaginationComponent correctly", () => {
    const { container } = render(
      <PaginationComponent
        pages={mockPages}
        currentPage={mockCurrentPage}
        handlePrevClick={mockHandlePrevClick}
        handleNextClick={mockHandleNextClick}
        handlePageClick={mockHandlePageClick}
        startPage={mockStartPage}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("calls handlePrevClick on clicking Previous button", () => {
    const { getByText } = render(
      <PaginationComponent
        pages={mockPages}
        currentPage={mockCurrentPage}
        handlePrevClick={mockHandlePrevClick}
        handleNextClick={mockHandleNextClick}
        handlePageClick={mockHandlePageClick}
        startPage={mockStartPage}
      />
    );

    fireEvent.click(getByText("Previous"));

    expect(mockHandlePrevClick).toHaveBeenCalled();
  });

  it("calls handleNextClick on clicking Next button", () => {
    const { getByText } = render(
      <PaginationComponent
        pages={mockPages}
        currentPage={mockCurrentPage}
        handlePrevClick={mockHandlePrevClick}
        handleNextClick={mockHandleNextClick}
        handlePageClick={mockHandlePageClick}
        startPage={mockStartPage}
      />
    );

    fireEvent.click(getByText("Next"));

    expect(mockHandleNextClick).toHaveBeenCalled();
  });

  it("calls handlePageClick with the correct page number", () => {
    const { getByText } = render(
      <PaginationComponent
        pages={mockPages}
        currentPage={mockCurrentPage}
        handlePrevClick={mockHandlePrevClick}
        handleNextClick={mockHandleNextClick}
        handlePageClick={mockHandlePageClick}
        startPage={mockStartPage}
      />
    );

    fireEvent.click(getByText("2")); // Assuming there is a '2' in the pagination

    expect(mockHandlePageClick).toHaveBeenCalledWith(2);
  });
});
