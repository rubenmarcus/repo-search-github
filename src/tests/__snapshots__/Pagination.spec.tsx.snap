// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationComponent renders the PaginationComponent correctly 1`] = `
<div>
  <nav
    aria-label="pagination"
    class="mx-auto flex w-full justify-center my-5"
    role="navigation"
  >
    <ul
      class="flex flex-row items-center gap-1"
    >
      <li
        class=""
      >
        <li
          class=""
        >
          <a
            aria-label="Go to previous page"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 gap-1 pl-2.5"
            href="#"
          >
            <svg
              class="lucide lucide-chevron-left h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m15 18-6-6 6-6"
              />
            </svg>
            <span>
              Previous
            </span>
          </a>
        </li>
      </li>
      <li
        class=""
      >
        <li
          class=""
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10 bg-white text-black"
            href="#"
          >
            1
          </a>
        </li>
         
      </li>
      <li
        class=""
      >
        <li
          class=""
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
            href="#"
          >
            2
          </a>
        </li>
         
      </li>
      <li
        class=""
      >
        <li
          class=""
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
            href="#"
          >
            3
          </a>
        </li>
         
      </li>
      <li
        class=""
      >
        <li
          class=""
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
            href="#"
          >
            4
          </a>
        </li>
         
      </li>
      <li
        class=""
      >
        <li
          class=""
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
            href="#"
          >
            5
          </a>
        </li>
         
      </li>
      <li
        class=""
      >
        <span
          aria-hidden="true"
          class="flex h-9 w-9 items-center justify-center"
        >
          <svg
            class="lucide lucide-more-horizontal h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="1"
            />
            <circle
              cx="19"
              cy="12"
              r="1"
            />
            <circle
              cx="5"
              cy="12"
              r="1"
            />
          </svg>
          <span
            class="sr-only"
          >
            More pages
          </span>
        </span>
      </li>
      <li
        class=""
      >
        <li
          class=""
        >
          <a
            aria-label="Go to next page"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 gap-1 pr-2.5"
            href="#"
          >
            <span>
              Next
            </span>
            <svg
              class="lucide lucide-chevron-right h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m9 18 6-6-6-6"
              />
            </svg>
          </a>
        </li>
      </li>
    </ul>
  </nav>
</div>
`;
